packages/tokenbank_v1/contracts/modules/security/RecoveryModule.sol

```diff
@@ -0,0 +1,164 @@
+        require(signers.length >= (guardianCount + 1)/2, "NOT_ENOUGH_SIGNER");
+        require(isWalletOwnerOrGuardian(wallet, signers), "UNAUTHORIZED");
+
+        recovery.newOwner = newOwner;
+        recovery.completeAfter = now + recoveryPeriod;
+        recovery.guardianCount = guardianCount;
+
+        securityStorage.setLock(wallet, now + lockPeriod);
+
+        emit RecoveryStarted(wallet, newOwner, recovery.completeAfter);
+    }
+
+    function cancelRecovery(
+        address   wallet,
+        address[] calldata signers
+        )
+        external
+        nonReentrantExceptFromThis
+        onlyFromMetaTx
+    {
+        WalletRecovery storage recovery = wallets[wallet];
+        require(recovery.completeAfter > 0, "NOT_STARTED");
+
+        uint guardianCount = wallets[wallet].guardianCount;
+        require(signers.length >= (guardianCount + 1) / 2, "NOT_ENOUGH_SIGNER");
+        require(isWalletOwnerOrGuardian(wallet, signers), "UNAUTHORIZED");
+
+        delete wallets[wallet];
+        securityStorage.setLock(wallet, 0);
+
+        emit RecoveryCancelled(wallet);
+    }
+
+    function completeRecovery(address wallet)
+        external
+        nonReentrantExceptFromThis
+        onlyFromMetaTx
+    {
+        WalletRecovery storage recovery = wallets[wallet];
+        require(recovery.completeAfter > 0, "NOT_STARTED");
+        require(recovery.completeAfter < now, "TWO_EARLY");
+
+        Wallet(wallet).setOwner(recovery.newOwner);
+
+        delete wallets[wallet];
+        securityStorage.setLock(wallet, 0);
+
+        emit RecoveryCompleted(wallet, recovery.newOwner);
+    }
+
+    function extractMetaTxSigners(
+        address /*wallet*/,
+        bytes4  method,
+        bytes   memory data
+        )
+        internal
+        view
+        returns (address[] memory signers)
+    {
+        uint skipBytes = 0;
+        if (method == this.startRecovery.selector) {
+            // data layout: {length:32}{sig:4}{_wallet:32}{_newOwner:32}{signer1:32}{signer2:32}{...}
+            skipBytes = 100;
+        } else if (method == this.cancelRecovery.selector) {
+            // data layout: {length:32}{sig:4}{_wallet:32}{signer1:32}{signer2:32}{...}
+            skipBytes = 68;
+        } else if (method == this.completeRecovery.selector){
+            // data layout: {length:32}{sig:4}{_wallet:32}
+            skipBytes = 68;
+            require(data.length == skipBytes, "INVALID_DATA");
+        }
+
+        require(
+            data.length >= skipBytes && (data.length - skipBytes) % 32 == 0,
+            "DATA_INVALID"
+        );
+
+        uint numSigners = (data.length - skipBytes) / 32;
+        signers = new address[](numSigners);
+
+        address signer;
+        for (uint i = 0; i < numSigners; i++) {
+            uint start = skipBytes + 32 * i;
+            assembly {signer := mload(add(data, start)) }
+            signers[i] = signer;
+        }
+    }
+}
```

Comment:
The `length` is the array length, right? Then, for example, with data like this:
```
{length:32}{sig:4}{_wallet:32}{_newOwner:32}{signer1:32}
```
`data.length == 100` and **not** `data.length == 132`, because the number of bytes used by `length` at the start is just some internal data for the array. But for reading the data using assembly you do need to use a correct offset that takes those `length` bytes into account.

Probably a good idea to add a utility method in `BytesUtil.sol` to read addresses stored in 32 bytes (there's already a 20 bytes one, but the current function names are pretty bad, so some renaming is a good idea anyway):
```
// Can be used to extract an address packed in 20 bytes
function readPackedAddress(
        bytes memory b,
        uint  offset
        )
        internal
        pure
        returns (address)
{
    return address(bytesToUintX(b, offset, 20) & 0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF);
}

// Can be used to extract an address packed in 32 bytes
function readAddress(
        bytes memory b,
        uint  offset
        )
        internal
        pure
        returns (address)
{
    return address(bytesToUintX(b, offset, 32) & 0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF);
}
```

This way we don't have to think about it:
```
function extractMetaTxSigners(
    address /*wallet*/,
    bytes4  method,
    bytes   memory data
    )
    internal
    view
    returns (address[] memory signers)
{
    uint skipBytes = 0;
    if (method == this.startRecovery.selector) {
        // data layout: {sig:4}{_wallet:32}{_newOwner:32}{signer1:32}{signer2:32}{...}
        skipBytes = 68;
    } else if (method == this.cancelRecovery.selector) {
        // data layout: {sig:4}{_wallet:32}{signer1:32}{signer2:32}{...}
        skipBytes = 36;
    } else if (method == this.completeRecovery.selector){
        // data layout: {sig:4}{_wallet:32}
        skipBytes = 36;
        require(data.length == skipBytes, "INVALID_DATA");
    } else {
        revert("UNKNOWN_DATA");
    }

    uint numSigners = (data.length.sub(skipBytes)) / 32;
    signers = new address[](numSigners);
    for (uint i = 0; i < numSigners; i++) {
        signers[i] = data.readAddress(skipBytes + 32 * i);
    }
}
```

Reply:
Actually I like that we know all the details :)